### PR TITLE
feat: Filter networks by window location

### PR DESCRIPTION
**Pull request by stanleyjones**
# Description

This PR adds a network for the END Labs ledger and filters which networks are available based on the location reported by the browser.

- [x] Add "GBOP Ledger" as a network option with the endpoint `/api-end`
- [x] Rename "Manifest Network" to "Manifest Ledger"
- [x] Filter available networks by browser-reported location

Closes https://github.com/liftedinit/operations/issues/6 (but doesn't really)

# Testing / Breaking Changes

The change was tested manually and required modifying the existing E2E tests, as they were built with the assumption that only one network would be available.

**BREAKING CHANGE:** The "Manifest Network" will no longer be available from front-ends running Alberto that do not have "alberto" somewhere in the domain name. While this change is the goal of this ticket, it also constitutes a breaking change. We can put it back once ledger clusters have static URLs for their APIs.

# Screenshots

**AS ALBERTO**

Screenshot 2022-12-05 at 2 28 05 PM

**AS WALLET.END-LABS**

Screenshot 2022-12-05 at 2 29 05 PM

### CODE REVIEW

File: config-overrides.js [hunk 1]
```diff
@@ -1,9 +1,17 @@
...
-  return webpackConfig;
-};
+    return webpackConfig
+  },
+  devServer: function (configDevServer) {
+    return function (proxy, allowedHosts) {
+      const config = configDevServer(proxy, allowedHosts)
+      return { ...config, public: "alberto.local" }
+    }
+  },
```
Comment (stanleyjones) [inline config-overrides.js]: This bit allows Webpack to serve Alberto locally at a specific domain (e.g. `https://alberto.local:3000`) provided you also add the following line to your `/etc/hosts`:

```
127.0.0.1  alberto.local
127.0.0.1  wallet.end-labs.local
```

It's useful in local development to test whether the hostname detection is working.

File: src/features/network/components/network-menu/network-menu.tsx [hunk 2]
```diff
@@ -38,20 +38,11 @@ export function NetworkMenu() {
...
-      activeId: s.activeId,
-      activeNetwork: s.byId.get(s.activeId),
-      setActiveId: s.setActiveId,
-    }),
-  )
+  const { activeNetwork, networks, setActiveId } = useNetworkStore(s => ({
+    networks: s.getNetworks(),
+    activeNetwork: s.getActiveNetwork(),
+    setActiveId: s.setActiveId,
+  }))
```
Comment (stanleyjones) [inline src/features/network/components/network-menu/network-menu.tsx:45]: All of this business logic was moved out of the view and into the state store. See below.

File: src/features/network/store.ts [hunk 3]
```diff
@@ -5,12 +5,16 @@ import { replacer, reviver } from "shared/helpers"
 import { NetworkId, NetworkParams, NetworksState } from "./types"
 
 const initialState = {
-  activeId: 0,
-  nextId: 1,
-  byId: new Map([[0, { name: "Manifest", url: "/api" }]]),
+  nextId: 2,
+  byId: new Map([
+    [0, { name: "Manifest Ledger", url: "/api", filter: "alberto" }],
+    [1, { name: "GBOP Ledger", url: "/api-end", filter: "end-labs" }],
```
Comment (stanleyjones) [inline src/features/network/store.ts:12]: @jgryffindor Please double-check the `url` here. I'm pretty sure it's what I heard you say. Easy to change.

File: src/features/network/store.ts [hunk 4]
```diff
@@ -19,8 +23,22 @@ interface NetworkActions {
...
+    (set, get) => ({
       ...initialState,
+      getNetworks: () => {
+        const hostname = window.location.hostname
+        const networks = Array.from(get().byId)
+          .map(([id, network]) => ({ id, ...network }))
+          .sort(({ name: a }, { name: b }) =>
+            a.toLowerCase() < b.toLowerCase() ? -1 : 1,
+          )
+          .filter(({ filter }) => !filter || hostname.includes(filter))
```
Comment (stanleyjones) [inline src/features/network/store.ts]: The mapping, sorting, and filtering has been moved here from the view.

File: config-overrides.js [hunk 5]
```diff
@@ -1,9 +1,18 @@
...
 
-  return webpackConfig;
-};
+    return webpackConfig
+  },
+  devServer: function (configDevServer) {
+    return function (proxy, allowedHosts) {
+      const config = configDevServer(proxy, allowedHosts)
+      return config
+      // return { ...config, public: "wallet.end-labs.local" }
```
Comment (stanleyjones) [inline config-overrides.js:15]: Comment out L14 and uncomment L15 to serve the app from a different domain name during local development. Make sure to add that domain in your `/etc/hosts` as well (e.g. `127.0.0.1  wallet.end-labs.local`).

File: e2e/tests/home.spec.ts [hunk 6]
```diff
@@ -95,10 +95,11 @@ test("remove a network", async ({ page }) => {
...
   const removeBtn = page.locator("data-testid=remove network button")
   expect(removeBtn).toBeDisabled()
+  const name = await page.locator('input[name="name"]').inputValue()
   const url = await page.locator('input[name="url"]').inputValue()
   await page.locator('input[name="deleteUrl"]').fill(url as string)
   expect(removeBtn).not.toBeDisabled()
   await removeBtn.click()
   const networkName = await homePage.networkMenuTrigger.textContent()
-  expect(networkName).toBe("no network selected")
+  expect(networkName).not.toBe(name)
```
Comment (stanleyjones) [inline e2e/tests/home.spec.ts:104]: The test assumed that no network would be available after deleting one. I changed it to confirm that the deleted network was no longer active.

File: src/features/network/store.ts [hunk 7]
```diff
@@ -19,8 +24,28 @@ interface NetworkActions {
...
+          .map(([id, network]) => ({ id, ...network }))
+          .sort(({ name: a }, { name: b }) =>
+            a.toLowerCase() < b.toLowerCase() ? -1 : 1,
+          )
+          .filter(
+            ({ filter }) =>
+              !filter ||
+              hostname.includes(filter) ||
+              hostname.includes("localhost"),
+          )
```
Comment (stanleyjones) [inline src/features/network/store.ts:41]: All the mapping, sorting, and filtering from the view has been moved here. Note that _all_ networks are available if you're just viewing from `localhost` — but see above for a trick to set your domain to something else in local development.

File: src/features/network/store.ts [hunk 8]
```diff
@@ -19,8 +24,28 @@ interface NetworkActions {
...
+        return networks
+      },
+      getActiveNetwork: () => {
+        const networks = get().getNetworks()
+        const activeNetwork = networks.find(({ id }) => id === get().activeId)
+        return activeNetwork ?? networks[0]
```
Comment (stanleyjones) [inline src/features/network/store.ts:47]: It's possible that the "active network" in your browser cache no longer exists, in which case fall back to the first one alphabetically.

File: src/features/network/store.ts (hunk 3 above)
Comment (jgryffindor) [reply src/features/network/store.ts:12]: Looks good!

**Review by jgryffindor — APPROVED** [review]

**Review by mjsorribas — APPROVED** [review]

**Pull request merged**